presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteConstantArrayContainsToInExpression.java

```diff
@@ -0,0 +1,98 @@
+import com.facebook.presto.common.type.ArrayType;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.expressions.RowExpressionRewriter;
+import com.facebook.presto.expressions.RowExpressionTreeRewriter;
+import com.facebook.presto.metadata.FunctionAndTypeManager;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.relational.FunctionResolution;
+import com.google.common.collect.ImmutableList;
+
+import static com.facebook.presto.SystemSessionProperties.isRwriteConstantArrayContainsToInExpressionEnabled;
+import static com.facebook.presto.common.type.TypeUtils.readNativeValue;
+import static com.facebook.presto.sql.relational.Expressions.constant;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class RewriteConstantArrayContainsToInExpression
+        extends RowExpressionRewriteRuleSet
+{
+    public RewriteConstantArrayContainsToInExpression(FunctionAndTypeManager functionAndTypeManager)
+    {
+        super(new Rewriter(functionAndTypeManager));
+    }
+
+    @Override
+    public boolean isRewriterEnabled(Session session)
+    {
+        return isRwriteConstantArrayContainsToInExpressionEnabled(session);
+    }
+
+    private static class Rewriter
+            implements PlanRowExpressionRewriter
+    {
+        private final ContainsToInRewriter caseToMapRewriter;
+
+        public Rewriter(FunctionAndTypeManager functionAndTypeManager)
+        {
+            requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+            this.caseToMapRewriter = new ContainsToInRewriter(functionAndTypeManager);
+        }
+
+        @Override
+        public RowExpression rewrite(RowExpression expression, Rule.Context context)
+        {
+            return RowExpressionTreeRewriter.rewriteWith(caseToMapRewriter, expression);
+        }
+    }
+
+    private static class ContainsToInRewriter
+            extends RowExpressionRewriter<Void>
+    {
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final FunctionResolution functionResolution;
+
+        private ContainsToInRewriter(FunctionAndTypeManager functionAndTypeManager)
+        {
+            this.functionAndTypeManager = functionAndTypeManager;
+            this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+        }
+
+        @Override
+        public RowExpression rewriteCall(CallExpression node, Void context, RowExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (functionResolution.isArrayContainsFunction(node.getFunctionHandle()) && node.getArguments().get(0) instanceof ConstantExpression) {
```

Comment:
Maybe flip the order - micro optimization :)

Also I wonder if we should write an iterative rule applying to only project/filter

Reply:
> Maybe flip the order - micro optimization :)

We need to add additional check that the function has at least one argument then, maybe not worth it?

> Also I wonder if we should write an iterative rule applying to only project/filter

Sure, Let me override the rules() function to only return filter and project rules.

Reply:
yes you are right about the optimization

About filter/project - just trying to make sure we apply only when there is a chance